docs/context_locals.rst

```diff
@@ -0,0 +1,375 @@
+
+Context locals
+================================================================================
+
+Context local objects ("context locals") are global objects that manage data
+specific to the current greenlet or thread ("the context"). Flask uses
+context locals, like ``request``, ``session``, ``current_app``, and ``g``, for
+data related to the current request ("the request context"). These objects are
+only bound when the application is processing a request.
+
+Motivation
+--------------------------------------------------------------------------------
+
+Many web frameworks, like Django, call view functions with data specific to the
+current request ("the request context"). This is in order to stay thread-safe;
+in these frameworks, view functions generate a response by passing the request
+context throughout their logic.
+
+The problem with this approach is that it can make application logic verbose,
+opaque, and expensive to maintain. Since dependency is transitive, many
+functions will depend on the request context only because they depend on some
+function which depends on the request context. [2]_ This is especially
+frustrating when a dependency on the request context is discovered or eliminated
+in a deeply nested function that causes the signature of all of its callers and
+their callers to be rewritten. This explains why some Django modules fetch the
+request context themselves. [*]_
+
+Many frameworks solve these problems by making the request context globally
+available. However, globals introduce three new problems. First, they can make
+large applications unmaintainable. Second, they aren't thread safe. Third, even
+if we can solve the first two problems, the behavior of thread-safe global is
+complex.
+
+Flask aims to make it quick and easy to write a traditional web application.
+[1]_ So, while globals can make a large application hard to maintain, Flask
+considers this out of the scope of the project. (Further, with responsible use
+it should be possible to tame the complexities of these globals; they do not
+manage state and are all singletons-- that is, there is one ``request`` per
+request).
+
+The standard way to make globals thread-safe is to use thread local storage,
+which Python has supported via ``threading.local()`` since 2.4. However,
+thread locals are not viable in web applications for two reasons. First, WSGI
+does not guarantee that every request will get its own thread; web servers may
+reuse threads for requests, which could pollute the thread local object and leak
+memory. Second, some popular web servers, like Gunicorn, handle concurrency
+without threads. Flask solves both of these problems by taking advantage of
+context local objects from Werkzeug; global objects that manage data specific to
+the current request.
+
+Thread-safe globals do introduce complexity into Flask. Developers who are
```

Comment:
This first sentence feels a little awkward. It uses the word `do` to imply a negative condition to a previously mentioned benefit, but doesn't explicitly point to the benefit. It might read easier if you add a "however" or get rid of the "do" in the sentence.